crates/circuit/src/dag_circuit.rs

```diff
@@ -5785,6 +5785,26 @@ impl DAGCircuit {
         }
     }
 
+    // Filter any nodes that don't match a given predicate function
+    pub fn filter_op_nodes<F, E>(&mut self, mut predicate: F) -> Result<(), E>
```

Comment:
I think the term "filter" makes more sense in the context of a transpiler pass / pipeline than it does directly on `DAGCircuit`, but I'm not sure I can think of a better name. Maybe something like `remove_op_nodes_by_filter`.

I'm fine with this comment being closed if anyone (yourself included) prefers the current name.

Reply:
Personally I thought this was pretty clear, using filter as a verb typically means to remove like this. But I'll let others chime in here, because my thoughts/reading on things aren't always clear to others.

Reply:
I think your thoughts on it are valid. The name just makes me think of an iterator kind of situation i.e. if you gave me just the method name I would not guess that it took a `&mut self`. But, it does. So it's up to personal preference I suppose.

Reply:
fwiw, compare Rust's `Iterator::filter` (new iterator containing only matching items) and Python's builtin `filter` (same).  Changing to `remove_op_nodes_by_filter` would invert the polarity of the `predicate` compared to the existing code and would probably conflict with Rust and Python builtins.

Reply:
and for the `&self` vs `&mut self` point, both Rust and Python's methods are mutations on the incoming iterator, which is consistent with `&mut self` (well, strictly `self`, but the analogy starts breaking down there

Reply:
Heh, the analogy breaking down there is the problem I have with this, since we're not consuming the DAG, so it's no longer a filter of some pipelined data flow, but instead removals from a sitting pool.

But, I'm fine with keeping the name as is since I'm clearly in the minority on this 🙂.